tensorflow/core/platform/logging_test.cc

```diff
@@ -97,3 +97,127 @@ TEST(InternalLogString, Basic) {
 }
 
 }  // namespace tensorflow
+
+#if defined(__linux__) && !defined(PLATFORM_POSIX_ANDROID)
+
+#include "absl/strings/str_split.h"
+#include <vector>
+#include <sstream>
+
+#define _GNU_SOURCE
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+namespace tensorflow {
+
+class LogSinkTest : public ::testing::Test {
+ protected:
+  static const size_t kBufLen = 1024;
+
+  void SetUp() override {
+    ASSERT_EQ(pipe(out_pipe_), 0);
+
+    auto flags = fcntl(out_pipe_[0], F_GETFL); 
+    flags |= O_NONBLOCK; 
+    fcntl(out_pipe_[0], F_SETFL, flags);
+
+    saved_stderr_ = dup(STDERR_FILENO);
+    ASSERT_NE(saved_stderr_, -1);
+
+    ASSERT_NE(dup2(out_pipe_[1], STDERR_FILENO), -1);
+    ASSERT_EQ(close(out_pipe_[1]), 0);
+
+    auto sinks = TFGetLogSinks();
+#ifndef NO_DEFAULT_LOGGER
+    ASSERT_EQ(sinks.size(), 1);
+#else
+    ASSERT_EQ(sinks.size(), 0);
+#endif
+
+    for(auto sink : TFGetLogSinks()) {
+      TFRemoveLogSink(sink);
+    }
+
+    TFAddLogSink(&default_sink_);
+  }
+
+  void TearDown() override {
+    ASSERT_NE(dup2(saved_stderr_, STDERR_FILENO), -1);
+    saved_stderr_ = out_pipe_[0] = out_pipe_[1] = -1;
+
+    for(auto sink : TFGetLogSinks()) {
+      TFRemoveLogSink(sink);
+    }
+#ifndef NO_DEFAULT_LOGGER
+    TFAddLogSink(&default_sink_);
+#endif
+  }
+
+  std::string GetStdErr() {
+    fflush(stderr);
+
+    ssize_t len = read(out_pipe_[0], (void*)buf_, kBufLen);
+    
+    std::string ret;
+    if(len > 0) {
+      ret.assign(buf_, buf_ + len);
+    }
+    
+    return ret;
+  }
+
+ private:
+  int saved_stderr_ = -1;
+  int out_pipe_[2] = { -1, -1 };
+  char buf_[kBufLen + 1] = { '/0' };
+  static TFDefaultLogSink default_sink_;
+};
+
+TFDefaultLogSink LogSinkTest::default_sink_;
+
+struct TestSink : public TFLogSink {
```

Comment:
nit: use class and add a member function that returns ss.str()

Reply:
In a test, really? That seems like an overkill. :shrug: I guess I'll pretty it up.